proposals/42-human-readable-mutable-names.md

```diff
@@ -0,0 +1,235 @@
+
+#### What does done look like?
+_What specific deliverables should completed to consider this project done?_
+
+- DNS over HTTPS is supported in js/go-ipfs/libp2p
+    - User agency around DNS is respected: every code path doing a DNS lookup (resolving DNSLink, `/dnsaddr`, `/dns*` multiaddrs) should use either explicit or implicit DNS resolvers defined in IPFS node configuration file.
+      - By default OS-level DNS resolver is used as catch-all, but the user can delegate resolution of all or specific TLDs to specific DoH resolvers.
+      - Node tries to resolve with the most specific resolver (if present for TLD), and then fall back to the global one.
+    - Visual Aid / config mock-up
+      ```json
+        "DNS": {
+          "Resolvers": {
+            "*.eth": "https://eth.link-or-cloudflare/path/to/dns-query",
+            "*.crypto": "https://unstoppablesomething.example.com/path/to/dns-query",
+            "*.libre": "https://www.opennic.org/path/to/their/dns-query", 
+            "*": "https://mozilla.cloudflare-dns.com/dns-query"
+          }
+        }
+      ```
+- DNSLink provides interop for human-readable names and enables independence from PKI and ICANN.
+    - Anyone can set up and run own naming system without our involvement, blessing or endorsement.
+    - No vendor-specific client is included. 
+- IPNS provides a reliable solution for cryptographically-verifiable mutable names that won't break, as long someone is re-publishing the latest record.
+    - IPNS re-publishing works seamlessly, out of the box, just like providing CIDs for blocks in local datastore (pinned or not).
+    - It is possible to pin IPNS address to keep it around and ensure it resolves even when original publisher is gone.
+        - Node follows IPNS record updates and keeps re-publishing latest one if original author disappears.
+        - Node follows content updates behind the IPNS record, and update content pin every time IPNS record changes. 
+            - It is possible to opt-out from this behavior to re-publish IPNS record for `en.wikipedia-on-ipfs.org` without pinning all the data.
+
+
+####  What does success look like?
+_Success means impact. How will we know we did the right thing?_
+
+<!--
+Provide success criteria. These might include particular metrics, desired changes in the types of bug reports being filed, desired changes in qualitative user feedback (measured via surveys, etc), etc.
+-->
+
+- We see % of `dnslink=/ipns/` going up.
+- We see more IPNS records on DHT
+- We no longer see requests for supporting new naming systems.
+
+#### Counterpoints &amp; pre-mortem
+_Why might this project be lower impact than expected? How could this project fail to complete, or fail to be successful?_
+ 
+- Configurable per TLDs [DNS over HTTPS](https://en.wikipedia.org/wiki/DNS_over_HTTPS) resolver for  may be not enough to mitigate concerns around centralized nature of ICANN's DNS.
```

Comment:
Do you have any anecdotes that lead to think this could be the case?

Reply:
Cynical commentary goes like this:
- DoH by definition points at some centralized HTTPS server :trollface: 
- We don't want to ship some preconfigured DoH server for all ICANN names with go-ipfs, to avoid pointing all users at the same endpoint
  - DoH endpoint choice should be made by user or user agent
  - Still, it is possible that most people will use DoH endpoints from Google or Cloudflare, because those are fast and reliable, so we mitigated privacy leak and mitigated ISP-level spoofing, but then put all eggs in one basket